File: shiproom/external_validation/runner_image/patient-launcher.c
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

int main(int argc, char **argv) {
  if (argc < 2) return 64;
  pid_t child = fork();
  if (child < 0) return 65;
  if (child == 0) {
    if (setsid() < 0) _exit(66);
    execvp(argv[1], &argv[1]);
    perror("execvp");
    _exit(127);
  }
  int status;
  if (waitpid(child, &status, 0) < 0) return 67;
  /* The child is the dedicated session/process-group leader.  Background
     descendants cannot continue changing /output after its main command exits. */
  kill(-child, SIGTERM);
  struct timespec grace = { .tv_sec = 0, .tv_nsec = 250000000 };
  nanosleep(&grace, NULL);
  kill(-child, SIGKILL);
  if (WIFEXITED(status)) return WEXITSTATUS(status);
  return 128 + (WIFSIGNALED(status) ? WTERMSIG(status) : 1);
}

File: shiproom/external_validation/runner_image/patient_launcher.py
#!/usr/bin/env python3
"""Root-owned patient gateway: isolate every command in a new session."""
from __future__ import annotations
import os
import signal
import subprocess
import sys
import time

if len(sys.argv) < 2:
    raise SystemExit(64)

child = subprocess.Popen(sys.argv[1:], preexec_fn=os.setsid)
try:
    code = child.wait()
finally:
    # The command is the session leader.  Terminating the whole process group
    # prevents background children from writing after the main process exits.
    try: os.killpg(child.pid, signal.SIGTERM)
    except ProcessLookupError: pass
    time.sleep(0.25)
    try: os.killpg(child.pid, signal.SIGKILL)
    except ProcessLookupError: pass
raise SystemExit(code if code >= 0 else 128 + -code)
